ksql-engine/src/main/java/io/confluent/ksql/util/PersistentQueryMetadata.java

```diff
@@ -86,6 +91,9 @@ public PersistentQueryMetadata(
     this.materializationProvider =
         requireNonNull(materializationProvider, "materializationProvider");
     this.dataSourceType = Objects.requireNonNull(dataSourceType, "dataSourceType");
+    this.materialization = Suppliers.memoize(() -> materializationProvider
```

Comment:
Design note:  with this change the `PersistentQueryMetadata` now needs to know how to build the materialization.  This is adding more coupling to the code than is strictly necessary.  

A better approach might be to have the `PersistentQueryMetadata`  constructor accept the `Supplier<Optional<Materialization>>` rather than the current `Optional<MaterializationProvider>`.  This pushes the construction of the supplier to the calling code, (which already knows about pull query materialization), and decouples `PersistentQueryMetadata` from the pull query logic.  It returns to just being a dump pojo that just holds the data its passed.

However, given this is hopefully just a temporary fix to get ksqlDB over the line I'm happy for this to be merged as-is, as long as we have a github issue to track the longer term fix and we address that quickly after launch.  Please reference the github issue here for future reference.

Reply:
I have opened issues #3855 and #3854